host/sgx/loadelf.c

```diff
@@ -695,57 +728,170 @@ static oe_result_t _set_uint64_t_dynamic_symbol_value(
+                    (int64_t)(image->image_rva + symbol_definition.st_value) +
+                    addend;
+            }
+            /* Find the definition of the symbol in the dependent image */
+            else if (
+                elf64_find_dynamic_symbol_by_name(
+                    &dependency->elf, name, &symbol_definition) == 0 &&
+                symbol_definition.st_shndx != SHN_UNDEF)
+            {
+                p->r_addend =
+                    (int64_t)(
+                        dependency->image_rva + symbol_definition.st_value) +
+                    addend;
+            }
+            else
+            {
+                OE_TRACE_WARNING("symbol %s not found\n", name);
+            }
+        }
+        /* Patch non-symbolic relocation records */
+        else if (reloc_type == R_X86_64_RELATIVE)
+        {
+            p->r_addend += image->image_rva;
+        }
+    }
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
 static oe_result_t _patch_elf_image(
     oe_enclave_elf_image_t* image,
+    oe_enclave_elf_image_t* module_image,
     size_t enclave_size,
     size_t tls_page_count)
 {
     oe_result_t result = OE_UNEXPECTED;
     oe_sgx_enclave_properties_t* oeprops;
+    oe_enclave_module_info_t* module_info;
     uint64_t enclave_rva = 0;
 
     oeprops =
         (oe_sgx_enclave_properties_t*)(image->image_base + image->oeinfo_rva);
 
     assert((image->image_size & (OE_PAGE_SIZE - 1)) == 0);
-    assert((image->reloc_size & (OE_PAGE_SIZE - 1)) == 0);
+    if (module_image)
+        assert((module_image->image_size & (OE_PAGE_SIZE - 1)) == 0);
     assert((enclave_size & (OE_PAGE_SIZE - 1)) == 0);
 
-    /* Clear certain ELF header fields */
-    for (size_t i = 0; i < image->num_segments; i++)
-    {
-        const oe_elf_segment_t* seg = &image->segments[i];
-        elf64_ehdr_t* ehdr = (elf64_ehdr_t*)(image->image_base + seg->vaddr);
-
-        if (elf64_test_header(ehdr) == 0)
-        {
-            ehdr->e_shoff = 0;
-            ehdr->e_shnum = 0;
-            ehdr->e_shstrndx = 0;
-            break;
-        }
-    }
-
     oeprops->image_info.enclave_size = enclave_size;
     oeprops->image_info.oeinfo_rva = image->oeinfo_rva;
     oeprops->image_info.oeinfo_size = sizeof(oe_sgx_enclave_properties_t);
 
-    /* Set _enclave_rva to its own rva offset*/
+    /* Set _enclave_rva to its own rva offset */
     OE_CHECK(_get_dynamic_symbol_rva(image, "_enclave_rva", &enclave_rva));
     OE_CHECK(
         _set_uint64_t_dynamic_symbol_value(image, "_enclave_rva", enclave_rva));
 
-    /* reloc right after image */
+    /* reloc right after image
+     * Note that the image_info holds the non-padded size. */
     oeprops->image_info.reloc_rva = image->image_size;
     oeprops->image_info.reloc_size = image->reloc_size;
+    if (module_image)
+    {
+        oeprops->image_info.reloc_rva += module_image->image_size;
+        oeprops->image_info.reloc_size += module_image->reloc_size;
```

Comment:
This is the value prior to page alignment though, shouldn't this be patching to the page-aligned size?

Reply:
I updated the PR and make the relocation patching (merging + padding) right after loading the image. This allows for aligning more with existing behavior.